File: ios/SSHChat/PtyNoise.swift
import Foundation

enum PtyNoise {
    private static let timePrefix = try! NSRegularExpression(pattern: #"^\[\d{1,2}:\d{2}(?::\d{2})?]\s*"#)

    /// Drop prompt_toolkit / PTY echo noise (same idea as Android PtyNoise).
    static func shouldDrop(_ raw: String) -> Bool {
        var t = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if t.isEmpty { return true }

        // Real chat/system lines always carry a bracket tag — never drop those.
        // Keeps [#room] / [*] / [user] / [OK] visible even if prompt echo heuristics misfire.
        if t.contains("[#") || t.hasPrefix("[*]") || t.hasPrefix("[OK]") || t.hasPrefix("[ERROR]")
            || t.hasPrefix("[INFO]") || t.hasPrefix("[+]") || t.hasPrefix("[-]") || t.hasPrefix("[!]")
        {
            return false
        }
        // "[12:34] [#room] ..." after optional local time — still chat.
        if t.range(of: #"^\[\d{1,2}:\d{2}"#, options: .regularExpression) != nil {
            return false
        }

        while true {
            let range = NSRange(t.startIndex..., in: t)
            guard let match = timePrefix.firstMatch(in: t, range: range),
                  let matchRange = Range(match.range, in: t)
            else { break }
            t = String(t[matchRange.upperBound...]).trimmingCharacters(in: .whitespaces)
        }

        if t.hasPrefix("?[") && (t.hasSuffix("A") || t.hasSuffix("K")) { return true }
        if t.hasPrefix("WARNING: your terminal doesn't support cursor position requests") { return true }
        // "> test" commit/redraw echoes — not chat content.
        if t == ">" || t.hasPrefix("> ") { return true }
        let stripped = t.trimmingCharacters(in: .whitespaces)
        if stripped.hasPrefix(">") {
            // Keep only if the echo somehow still embeds a chat bracket.
            if !stripped.contains("[") { return true }
        }
        return false
    }
}

File: android/app/src/main/java/chat/ssh/sshchat/PtyNoise.kt
package chat.ssh.sshchat

/** Drop prompt_toolkit / PTY echo noise (same idea as sshchat_gui._should_drop_line). */
object PtyNoise {
    private val timePrefix = Regex("""^\[\d{1,2}:\d{2}(?::\d{2})?]\s*""")
    private val promptPrefix = Regex("""(?:^|\s)>+ +\s*""")

    fun shouldDrop(raw: String): Boolean {
        var t = raw.trim()
        if (t.isEmpty()) return true
        // Real chat/system tags — never drop (incl. CSI crumbs like `[K[*] 9 …`).
        if ("[#" in t || "[*]" in t || t.startsWith("[OK]") || t.startsWith("[ERROR]")
            || t.startsWith("[INFO]") || t.startsWith("[+]") || t.startsWith("[-]") || t.startsWith("[!]")
        ) {
            return false
        }
        while (true) {
            val nxt = timePrefix.replaceFirst(t, "").trimStart()
            if (nxt == t) break
            t = nxt
        }
        if (t.startsWith("?[") && (t.endsWith("A") || t.endsWith("K"))) return true
        if (t.startsWith("WARNING: your terminal doesn't support cursor position requests")) return true
        // "> test" commit/redraw echoes — not chat content.
        if (t == ">" || t.startsWith("> ")) return true
        val stripped = t.trim()
        if (stripped.startsWith(">")) {
            val tail = promptPrefix.replaceFirst(stripped, "").trim()
            if (tail.isEmpty() || "[" !in tail) return true
        }
        return false
    }
}
